ddtrace/debugging/_exception/replay.py

```diff
@@ -47,13 +48,70 @@
 FRAME_FILE_TAG = "_dd.debug.error.%d.file"
 FRAME_LINE_TAG = "_dd.debug.error.%d.line"
 
+EXCEPTION_IDENT_LIMIT = 3600.0  # 1 hour
+EXCEPTION_IDENT_LIMITER: t.Dict[int, HourGlass] = {}
+
+
+ExceptionChain = t.Deque[t.Tuple[BaseException, t.Optional[TracebackType]]]
+
+
+def exception_ident(exc: BaseException, tb: t.Optional[TracebackType]) -> int:
+    """Compute the identity of an exception.
+
+    We use the exception type and the traceback to generate a unique identifier
+    that we can use to identify the exception. This can be used to rate limit
+    the number of times we capture information of the same exception.
+    """
+    h = 0
+    _tb = tb
+    while _tb is not None:
+        frame = _tb.tb_frame
+        h = (h << 1) ^ (id(frame.f_code) << 4 | frame.f_lasti)
+        h &= 0xFFFFFFFFFFFFFFFF
+        _tb = _tb.tb_next
+    return (id(type(exc)) << 64) | h
+
+
+def exception_chain_ident(chain: ExceptionChain) -> int:
+    h = 0
+    for exc, tb in chain:
+        h = (h << 1) ^ exception_ident(exc, tb)
+        h &= 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
+    return h
+
+
+def limit_exception(chain: ExceptionChain) -> bool:
+    try:
+        exc_ident = exception_chain_ident(chain)
+        hg = EXCEPTION_IDENT_LIMITER.get(exc_ident)
+        if hg is None:
+            # We haven't seen this exception yet, or it's been evicted
+            hg = EXCEPTION_IDENT_LIMITER[exc_ident] = HourGlass(duration=EXCEPTION_IDENT_LIMIT)
+            hg.turn()
+            return False
+
+        if hg.trickling():
+            # We have seen this exception recently
+            return True
+
+        # We haven't seen this exception in a while
+        hg.turn()
+
+        return False
+    finally:
+        if len(EXCEPTION_IDENT_LIMITER) > 1024:
+            # We limit the number of exception identities we track to avoid
+            # memory leaks.
+            sorted_keys = sorted(EXCEPTION_IDENT_LIMITER, key=lambda k: EXCEPTION_IDENT_LIMITER[k])
```

Comment:
Shouldn't the order be based on started_at / end_at?

Reply:
I thought it made more sense to sort by `end_at`, as these would be the timers that go off first. When the interval is the same for all the timers, these are the "older" ones (and `start_at` would essentially be equivalent to `end_at` modulo a constant time shift). So this will evict the older timers, increasing the odds that we will re-capture an exception that we haven't seen recently (compared to more recent ones).

Reply:
I was confused about the implicit nature of this call, clarified offline, resolving.